src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs

```diff
@@ -2797,6 +2799,66 @@ public async Task ReceivingMessagesPreventsConnectionTimeoutFromOccuring()
             }
         }
 
+        internal class PipeReaderWrapper : PipeReader
+        {
+            private readonly PipeReader _originalPipeReader;
+            private bool _waitingForRead;
+
+            public PipeReaderWrapper(PipeReader pipeReader)
+            {
+                _originalPipeReader = pipeReader;
+            }
+
+            public override void AdvanceTo(SequencePosition consumed) =>
+                _originalPipeReader.AdvanceTo(consumed);
+
+            public override void AdvanceTo(SequencePosition consumed, SequencePosition examined) =>
+                _originalPipeReader.AdvanceTo(consumed, examined);
+
+            public override void CancelPendingRead() =>
+                _originalPipeReader.CancelPendingRead();
+
+            public override void Complete(Exception exception = null) =>
+                _originalPipeReader.Complete(exception);
+
+            public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+            {
+                _waitingForRead = true;
+                try
+                {
+                    return await _originalPipeReader.ReadAsync(cancellationToken);
+                }
+                finally
+                {
+                    _waitingForRead = false;
+                }
+            }
+
+            public override bool TryRead(out ReadResult result) =>
+                _originalPipeReader.TryRead(out result);
+
+            public bool IsWaitingForRead()
+            {
+                return _waitingForRead;
```

Comment:
Could you make the test pass faster by using a TCS to implement a WaitForRead method that returns a task that completes once a read starts?

Reply:
I thought about that but was worried I'd have to write a bunch of synchronization logic to make sure the TCS is reset properly. I'll try it out, the test logic would look a lot nicer if we used a TCS...